platform/default/include/mbgl/util/default_thread_pool.hpp

```diff
@@ -6,17 +6,40 @@
 #include <mutex>
 #include <queue>
 #include <thread>
+#include <functional>
 
 namespace mbgl {
 
+struct ThreadLifecycle {
+    using ThreadData = void*;
+    ThreadLifecycle() = default;
+    ThreadLifecycle(std::function<ThreadData()> onThreadCreated_,
+                    std::function<void(ThreadData)> onThreadDestroyed_) :
+            onThreadCreatedFn(onThreadCreated_),
+            onThreadDestroyedFn(onThreadDestroyed_) {}
+
+    ThreadData onThreadCreated() const {
+        return onThreadCreatedFn();
+    }
+
+    void onThreadDestroyed(ThreadData threadData_) const {
+        onThreadDestroyedFn(threadData_);
+    }
+
+private:
+    std::function<ThreadData()> onThreadCreatedFn = [] { return nullptr; };
+    std::function<void(ThreadData)> onThreadDestroyedFn = [](ThreadData) {};
+};
+
 class ThreadPool : public Scheduler {
 public:
-    ThreadPool(std::size_t count);
+    explicit ThreadPool(std::size_t count, ThreadLifecycle _lifecycle = ThreadLifecycle());
     ~ThreadPool() override;
 
     void schedule(std::weak_ptr<Mailbox>) override;
 
 private:
+    ThreadLifecycle lifecycle;
```

Comment:
I would pass it by value to the threads in the pool and avoid storing instance on the main thread.